mava/utils/networks_utils.py

```diff
@@ -0,0 +1,126 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of an mlp module with layer normalisation."""
+# Adapted from https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/nets/mlp.py
+
+from typing import Any, Callable, Iterable, Optional
+
+import haiku as hk
+import jax
+import jax.numpy as jnp
+
+
+class MLP_NORM(hk.Module):
+    """A multi-layer perceptron module with an option for layer normalisation."""
+
+    def __init__(
+        self,
+        output_sizes: Iterable[int],
+        w_init: Optional[hk.initializers.Initializer] = None,
+        b_init: Optional[hk.initializers.Initializer] = None,
+        with_bias: bool = True,
+        activation: Callable[[jnp.ndarray], jnp.ndarray] = jax.nn.relu,
+        activate_final: bool = False,
+        layer_norm: bool = False,
+        name: Optional[str] = None,
+    ):
+        """Constructs an MLP with layer normalisation (MLP_NORM).
+        Args:
+            output_sizes: Sequence of layer sizes.
+            w_init: Initializer for :class:`~haiku.Linear` weights.
+            b_init: Initializer for :class:`~haiku.Linear` bias. Must be ``None`` if
+            ``with_bias=False``.
+            with_bias: Whether or not to apply a bias in each layer.
+            activation: Activation function to apply between :class:`~haiku.Linear`
+            layers. Defaults to ReLU.
+            activate_final: Whether or not to activate the final layer of the MLP.
+            layer_norm: apply layer normalisation to the hidden MLP layers.
+            name: Optional name for this module.
+        Raises:
+            ValueError: If ``with_bias`` is ``False`` and ``b_init`` is not ``None``.
+        """
+        if not with_bias and b_init is not None:
+            raise ValueError("When with_bias=False b_init must not be set.")
+
+        super().__init__(name=name)
+        self.with_bias = with_bias
+        self.w_init = w_init
+        self.b_init = b_init
+        self.activation = activation
+        self.activate_final = activate_final
+        self.layer_norm = layer_norm
+        layers = []
+        norms = []
+        output_sizes = tuple(output_sizes)
+        for index, output_size in enumerate(output_sizes):
+            layers.append(
+                hk.Linear(
+                    output_size=output_size,
+                    w_init=w_init,
+                    b_init=b_init,
+                    with_bias=with_bias,
+                    name="linear_%d" % index,
+                )
+            )
+            norms.append(
```

Comment:
I think it would be ideal if you first determine whether layer norm is True or False before calculating self.norms. Therefore in the case of non normalization you don't need to pass by this calculation.

Reply:
Thanks I will do that.